raven/src/main/java/org/opencadc/raven/ProtocolsGenerator.java

```diff
@@ -366,6 +361,26 @@ List<Protocol> doPullFrom(URI artifactURI, Transfer transfer, String authToken)
                 }
             }
         }
+        if ((storageResolvers != null) && storageResolvers.containsKey(artifactURI.getScheme())) {
```

Comment:
After following the advice about the README, you will have a single StorageResolver here (or null). You will have to call `StorageResolver.toURL(URI)` and figure out if it worked by catching `IllegalArgumentException` (scheme not supported) and probably cxhecking for null (that API not well specified). You can't really use the `getScheme()` method because that's not well behaved for a delegating resolver (again, poor API we have lived with).

Second issue: you can't blindly assume that the URL returned by a resolver supports any specific security method; the API simply doesn't specify. The only thing a resolver can do is generate a URL that would work with the security method possible using credentials in the subject, so in principle it would be correct to assign  a security method using `AuthenticationUtil.getAuthMethodFromCredentials(Subject)`... but in any case instead of the loop below you should only generate one protocol with a URL from an external resolver.

Reply:
Haven't changed the functionality, because there's no way to pick the right protocol so IMO it's best to give the client the option of trying them until one succeeds a better option becomes available.